### PR TITLE
ENH: Relative encoder support (homing, saving)

### PR DESCRIPTION
**Pull request by ZLLentz**
Implements the following:
- 5 different homing strategies, accessible via the ~`.HOMF` and `.HOMR` fields~ `:PLC:bHomeCmd` PV
- position save/restore for motors with relative encoders

I'm going to do a writeup here and include much of the information that follows in the README.

To activate a homing routing, put 1 to the EPICS ~`.HOMF` or `.HOMR` fields. As far as the TwinCAT program is concerned, these both mean the same thing: "home using our internal configuration".~ `:PLC:bHomeCmd` PV.

To configure a homing routine, set the `fHomePosition` variable to the position to use after homing, and set the `nHomingMode` variable to pick a strategy. These strategies are stored in `ENUM_EpicsHomeCmd` enum. All homing motion strategies move towards their destinations using the "Homing Velocity (towards plc cam)" parameter , then off of the switch using the "Homing velocity (off plc cam)" parameter. The normal options are:
- `LOW_LIMIT`: Move to the low limit (backward) switch. Set home position at the first point we see as we leave the switch.
- `HIGH_LIMIT`: Move to the high limit (forward) switch. Set home position at the first point we see as we leave the switch.
- `HOME_VIA_LOW`: Move towards the low limit (backward) switch, seeking out the position where `bHome` is `TRUE`. If we reach the limit switch without finding `bHome`, reverse direction and try moving towards the high switch. Set home position to the point just above the `bHome` signal.
- `HOME_VIA_HIGH`: Move towards the high limit (forward) switch, seeking out the position where `bHome` is `TRUE`. If we reach the limit switch without finding `bHome`, reverse direction and try moving towards the low switch. Set home position to the point just below the `bHome` signal.

There are two additional special-case options:
- `ABSOLUTE_SET`: When we ask for a home, do not move the motor- simply declare the current position as "home". This is basically a manual homing routine.
- `None`: Do not home! Ever! This is actually the default value, and the only one "implemented" prior to this PR.

If the homing strategy is set to any value other than `None`, the library knows we have a motor that wants save/restore. In these cases, the last position of the motor will be saved on TwinCAT shutdown and restored on the first cycle of the PLC program. This prevents the location of your relatively encoded motor from being lost when the PLC is reconfigured.

### CODE REVIEW

**Review by klauer — APPROVED** [review]
Gave the logic a reasonably close read-through. It all looks good, in my opinion. Added some comments as I read through which do not necessarily require a response.

As far as limits and overtravel: I'd assume that the default deceleration settings and the homing velocity setting are required to be set safely regardless of whether the homing routine runs or not, so a special hook for deceleration is unnecessary.

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        Axis:=stMotionStage.Axis,
+        Execute:=bLoad,
+        Position:=fPosition);
+
+    // Only load once, at startup
+    bLoad R= fbSetPos.Done OR fbSetPos.Error;
+
+    // Keep the error latched, it can disappear if Execute is set to FALSE
+    IF fbSetPos.Error THEN
+        nLatchError := fbSetPos.ErrorID;
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:46]: Ah I completely forgot about this, I never saw position save/restore error a single time in testing. It deserves some handling as you describe.

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    END_IF
+
+    // Do not save if we're currently loading
+    IF NOT bLoad THEN
+        fPosition := stMotionStage.stAxisStatus.fActPosition;
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:51]: I can check if there is an NC error, and I know the error range that corresponds to NC errors. That might be the best-effort try here to avoid saving a bad value.

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        fbSetPos(
+            Axis:=stMotionStage.Axis,
+            Execute:=bExecute,
+            Position:=stMotionStage.fHomePosition);
+        bBusy := fbSetPos.Busy;
+        bDone := fbSetPos.Done;
+        bMove := FALSE;
+    ENUM_EpicsHomeCmd.NONE:
+        bMove := FALSE;
+        bBusy := NOT bExecute;
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU:80]: Ahh I probably swapped this around- I'll swap it back and double-check there isn't any funny business when you set homing to NONE

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+            END_IF
+        FINAL_SETPOS:
+            fbSetPos(
+                Axis:=stMotionStage.Axis,
+                Execute:=TRUE,
+                Position:=stMotionStage.fHomePosition);
+            nHomeStateMachine := IDLE;
+            bBusy := FALSE;
+            bDone := TRUE;
+        ERROR:
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU:229]: In testing I always ended up at a "normal-looking" position after any error handling, since the error state machine hops over to the final position set, which gives us the "home" position. The motor will not be allowed to move in this case because it shows as `bError`.

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            bError := TRUE;
+            nErrorID := fbJog.ErrorID;
+            nHomeStateMachine := FINAL_SETPOS;
+            fbSetPos(
+                Axis:=stMotionStage.Axis,
+                Execute:=FALSE);
+            IF bInterrupted THEN
+                stMotionStage.sCustomErrorMessage := 'Homing interrutped';
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU:237]: I actually saw this when working through the PR and totally forgot to fix it

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    IDLE: INT := 0;
+    NEXT_MOVE: INT := 1;
+    CHECK_FWD: INT := 2;
+    CHECK_BWD: INT := 3;
+    FINAL_MOVE: INT := 4;
+    FINAL_SETPOS: INT := 5;
+    ERROR: INT := 6;
+
+    FWD_START: LREAL := -99999999;
+    BWD_START: LREAL :=  99999999;
```
Comment (ZLLentz) [inline lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU:40]: Quick annotation here: this is a simpler way of disabling the soft limits that ends up being really obvious if something has gone wrong. If you turn the limits off/on, not only do you need to keep track of if you had soft limits set, but you need to always restore this properly or risk some issue.

Instead, I set position to a ridiculous value that can always move forward or backward. If this gets stuck for any reason it's very clear that something has gone wrong, rather than a silent failure of the soft limit marks.

Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU:40]: (this will become a comment in the code)

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    bLoad: BOOL;
+    nLatchError: UDINT;
+END_VAR
+VAR PERSISTENT
+    bSaved: BOOL;
+    fPosition: LREAL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[IF bEnable THEN
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:27]: Another mark on the "add a pre-commit check to start beckhoff files with a newline for easy github reviewing" list

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:47]: Addressed in https://github.com/pcdshub/lcls-twincat-motion/pull/108/commits/0db9b6c206514ff939c3ec7bc0603a570ef14296,
there should probably be a separate issue for logging all `DUT_MotionStatge` errors

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:60]: Addressed in https://github.com/pcdshub/lcls-twincat-motion/pull/108/commits/fca5cd0fa48999e252bb2a3af761c4ca283a3914

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU (hunk 7 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:28]: Linking to https://github.com/pcdshub/pre-commit-hooks/issues/10

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU]: Addressed in https://github.com/pcdshub/lcls-twincat-motion/pull/108/commits/4e82c91304d8d8a833fabc5da80fdc17b9027d27

File: lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_EncSaveRestore.TcPOU:48]: Linking to #79

**Comment by ZLLentz** [comment]
I thought this was going to be done, but it seems like the motor record HOMF/HOMR do some... odd things, like setting the VAL field, and ask for extra moves to specific positions. I didn't notice these until I started modifying pcdsdevices to get a typhos screen going. Rather than debug that rabbit hole I'm going to redo this a tiny bit to work off of an auxiliary PV to trigger the homing.

File: lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZLLentz) [reply lcls-twincat-motion/Library/POUs/Motion/FB_MotionHoming.TcPOU]: There was actually a small rabbit hole here- they needed to flash on/off in the end to make it work correctly, as in https://github.com/pcdshub/lcls-twincat-motion/pull/108/commits/93b3c48b1bed0f617b475fbf145c8fb23fa2f59d

**Comment by ZLLentz** [comment]
This now works very well in my test setup! There were a few remaining edge cases related to the motor record status. I've made the fixes so that now:
- DMOV and MOVN are set appropriately
- The VAL field updates after homing is complete
- The RBV field updates during a homing routine

**Comment by ZLLentz** [comment]
Any more thoughts before I merge? @klauer @slacAWallace @ghalym

**Review by klauer — APPROVED** [review]
Looking through the diffs, my original comments have been addressed. 
* Should make an issue regarding the HOMF/HOMR-based homing routines. 
* pcdsdevices will need updating for the new PVs (have to decide how we'll support these two versions)

Nothing else in the diff jumped out as concerning to me, and I think the fact that it works well is plenty good to me.

(@slacAWallace ? @ghalym ?)

**Comment by ZLLentz** [comment]
> have to decide how we'll support these two versions

Right, that's annoying. I was planning to update the LFE/KFE/TMO plcs, but that doesn't guarantee that everything is handled.

**Comment by ZLLentz** [comment]
I'm going to merge this now
